### PR TITLE
feature/M2-1345: add subscale section form

### PR DESCRIPTION
**Pull request by alex-i-e**

### CODE REVIEW

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  subscaleId: string,
+  items: ItemFormValues[] = [],
+  subscales: ActivitySettingsSubscale[] = [],
+) => {
+  if (!items) return [];
+
+  const itemsMap = getMapFromList(items);
+  const subscalesMap = getMapFromList(subscales);
+  const subscaleElements = subscales
+    .filter((subscale) => subscale.id !== subscaleId)
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:64]: use reduce

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+export const getItemElementName = (item: ItemFormValues) =>
+  `${getItemNameInSubscale(item)}: ${t(item.question)}`;
+
+export const getSubscaleElementName = (
+  subscale: ActivitySettingsSubscale,
+  subscalesMap: Record<string, ActivitySettingsSubscale>,
+  itemsMap: Record<string, ItemFormValues>,
+) =>
+  `${t('subscale')}: ${subscale.name} (${subscale.items
+    .map((itemId) => {
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:36]: pls use reduce

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+export const getPropertiesToFilterByIds = (
+  items: ItemFormValues[] = [],
+  subscales: ActivitySettingsSubscale[] = [],
+) => {
+  const itemsMap = getMapFromList(items);
+  const subscalesMap = getMapFromList(subscales);
+  const allSubscaleIds = subscales.map((subscale) => getEntityKey(subscale));
+  const allItemIds = items.map((item) => getEntityKey(item));
+  const mergedIds = allSubscaleIds.concat(allItemIds);
+  const usedUniqueElementsIds = [
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:86]: rename pls without use

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  };
+};
+
+export const getNotUsedElements = (
+  subscalesMap: Record<string, ActivitySettingsSubscale>,
+  itemsMap: Record<string, ItemFormValues>,
+  mergedIds: string[],
+  usedUniqueElementsIds: string[],
+) => {
+  const filteredIds = mergedIds.filter((id) => !usedUniqueElementsIds.includes(id));
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:104]: pls use reduce

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+export const getUsedWithinSubscalesElements = (
+  subscales: ActivitySettingsSubscale[] = [],
+  subscalesMap: Record<string, ActivitySettingsSubscale>,
+  itemsMap: Record<string, ItemFormValues>,
+  mergedIds: string[],
+  usedUniqueElementsIds: string[],
+) => {
+  const filteredIds = mergedIds.filter((id) => usedUniqueElementsIds.includes(id));
+  const elements = filteredIds
+    .map((id) => {
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:137]: same here

File: src/shared/components/DataTable/DataTable.tsx [hunk 6]
```diff
@@ -44,60 +47,62 @@ export const DataTable = ({
...
+                <TableCell sx={{ width: '2.8rem', backgroundColor: 'inherit' }}>
+                  <Checkbox
+                    checked={isAllSelected}
+                    onChange={handleSelectAll}
+                    disabled={!data?.length}
+                  />
+                </TableCell>
+              )}
+              {columns?.map(({ key, label }) => (
+                <TableCell sx={{ backgroundColor: 'inherit' }} key={`data-table-head-${key}`}>
```
Comment (kdanilenka32) [inline src/shared/components/DataTable/DataTable.tsx:65]: any ide not to use index for key

File: src/shared/utils/func.ts [hunk 7]
```diff
@@ -1 +1,15 @@
 export const falseReturnFunc = () => false;
+
+export const getEntityKey = <T extends { id?: string; key?: string }>(entity: T) =>
+  entity?.id ?? entity?.key ?? '';
+
+export const getMapFromList = <T extends { id?: string; key?: string }>(
```
Comment (kdanilenka32) [inline src/shared/utils/func.ts:6]: pls rename this, you not return Map

File: src/shared/utils/func.ts [hunk 8]
```diff
@@ -1 +1,15 @@
 export const falseReturnFunc = () => false;
```
Comment (kdanilenka32) [inline src/shared/utils/func.ts:1]: Pls rename file more informative

File: src/shared/components/DataTable/DataTable.tsx [hunk 9]
```diff
@@ -44,60 +47,62 @@ export const DataTable = ({
...
+          <TableBody>
+            {data?.map((item, index) => {
+              const isSelected = selected?.includes(item.id);
+
+              return (
+                <TableRow key={`data-table-row-${index}`}>
+                  {selectable && (
+                    <TableCell sx={{ width: '2.8rem', backgroundColor: 'inherit' }}>
+                      <Checkbox
+                        key={`data-table-cell-checkbox-${isSelected}`}
```
Comment (kdanilenka32) [inline src/shared/components/DataTable/DataTable.tsx:82]: no way to use this key

File: src/shared/components/DataTable/DataTable.styles.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { styled } from '@mui/material';
```
Comment (kdanilenka32) [inline src/shared/components/DataTable/DataTable.styles.ts:1]: space

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+  const itemsMap = getMapFromList(items);
+  const subscalesMap = getMapFromList(subscales);
+  const subscaleElements = subscales
+    .filter((subscale) => subscale.id !== subscaleId)
+    .map((subscale) => ({
+      id: subscale.id ?? '',
+      [SubscaleColumns.Name]: getSubscaleElementName(subscale, subscalesMap, itemsMap),
+    }));
+  const itemElements = items.map((item) => ({
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:69]: space

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.tsx [hunk 12]
```diff
@@ -1,6 +1,124 @@
...
+    control,
+    name: 'subscales',
+  });
+
+  const [subscales, calculateTotalScoreSwitch]: [
+    AppletFormValues['subscales'],
+    AppletFormValues['calculateTotalScoreSwitch'],
+  ] = useWatch({ name: ['subscales', 'calculateTotalScoreSwitch'] });
+  const { activity } = useCurrentActivity();
+  const filteredItems = (activity?.items ?? []).filter(filterItemElements);
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.tsx:43]: better || [] here

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscaleHeaderContent/SubscaleHeaderContent.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  const { remove: removeSubscale } = useFieldArray({
+    control,
+    name: 'subscales',
+  });
+
+  const handleOnLookupTable = () => false;
+
+  return (
+    <StyledFlexTopCenter>
+      {' '}
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscaleHeaderContent/SubscaleHeaderContent.tsx:20]: do y need   {' '} ?

File: src/modules/Builder/features/ActivitySettings/ActivitySettings.const.tsx [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import { ButtonPropsVariantOverrides } from '@mui/material';
+import { Svg } from 'shared/components';
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/ActivitySettings.const.tsx:2]: space

File: src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/ToggleItemContainer/ToggleItemContainer.types.ts [hunk 15]
```diff
@@ -1,10 +1,14 @@
...
+import { SxProps } from '@mui/material';
 
 export type SharedToggleItemProps = {
   open?: boolean;
 };
 export type ToggleItemProps = {
   title: string;
-  HeaderContent: FC<SharedToggleItemProps>;
-  Content: FC<SharedToggleItemProps>;
+  HeaderContent: FC<SharedToggleItemProps & any>;
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/ToggleItemContainer/ToggleItemContainer.types.ts:9]: any solutions to fix any mb Jsx.Element

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.const.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+import i18n from 'i18n';
+import { SubscaleTotalScore } from 'shared/consts';
+
+const { t } = i18n;
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.const.ts:4]: space pls

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.tsx [hunk 17]
```diff
@@ -1,6 +1,124 @@
...
+  const handleAddSubscale = () => {
+    appendSubscale(getSubscalesDefaults());
+  };
+
+  const handleOnTotalScoreLookupTable = () => false;
+
+  return (
+    <StyledButtonsContainer>
+      {subscales?.map((subscale, index) => (
+        <ToggleItemContainer
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.tsx:71]: key ?

File: src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+      const subscale = subscalesMap[itemId];
+      if (item) return getItemNameInSubscale(item);
+      if (subscale) return `${t('subscale')}: ${subscale.name}`;
+
+      return '';
+    })
+    .filter(Boolean)
+    .join(', ')})`;
+
+export const filterItemElements = (item: ItemFormValues) =>
```
Comment (kdanilenka32) [inline src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.utils.ts:47]: pls rename it returns boolean

File: src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/ToggleItemContainer/ToggleItemContainer.types.ts (hunk 15 above)
Comment (alex-i-e) [reply src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/ToggleItemContainer/ToggleItemContainer.types.ts]: have no idea

**Review by IhorNesterenko99 — APPROVED** [review]

**Review by kdanilenka32 — APPROVED** [review]

**Pull request merged**